# Association Rules

In [1]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder


from mlxtend.frequent_patterns import apriori, association_rules

from pyECLAT import ECLAT

In [2]:
cluster_1 = pd.read_csv('clusters/cluster_1.csv')
cluster_2 = pd.read_csv('clusters/cluster_2.csv')
cluster_3 = pd.read_csv('clusters/cluster_3.csv')
cluster_4 = pd.read_csv('clusters/cluster_4.csv')
cluster_5 = pd.read_csv('clusters/cluster_5.csv')
cluster_6 = pd.read_csv('clusters/cluster_6.csv')
cluster_7 = pd.read_csv('clusters/cluster_7.csv')
cluster_8 = pd.read_csv('clusters/cluster_8.csv')
cluster_9 = pd.read_csv('clusters/cluster_9.csv')

Adding a label 'cluster' to link each customer to each number of cluster and 'type_of_customer' to have the characteristics of the group that the customer belongs:

In [3]:
cluster_1['type_of_customer'] = 'Vegan People'
cluster_2['type_of_customer'] = 'High Spenders'
cluster_3['type_of_customer'] = 'Young Families with Issues'
cluster_4['type_of_customer'] = 'Loyal Stable Shoppers'
cluster_5['type_of_customer'] = 'Tech-Gamer Customers'
cluster_6['type_of_customer'] = 'University Low-Spenders'
cluster_7['type_of_customer'] = 'Hygiene Obsessed'
cluster_8['type_of_customer'] = 'Loyal Average Spenders'
cluster_9['type_of_customer'] = 'Big Family'

In [4]:
datasets = [cluster_1, cluster_2, cluster_3, cluster_4, cluster_5, cluster_6, cluster_7, cluster_8, cluster_9]

scored_customers = combined_dataset = pd.concat(datasets)
scored_customers.drop(columns=['kids_home', 'teens_home', 'number_complaints', 'distinct_stores_visited', 'lifetime_spend_groceries', 'lifetime_spend_electronics', 'typical_hour', 'lifetime_spend_vegetables', 'lifetime_spend_nonalcohol_drinks', 'lifetime_spend_alcohol_drinks', 'lifetime_spend_meat', 'lifetime_spend_fish', 'lifetime_spend_hygiene', 'lifetime_spend_videogames', 'lifetime_spend_petfood', 'lifetime_total_distinct_products', 'loyalty_card_number', 'customer_age', 'years_since_first_transaction'], inplace=True)
scored_customers.to_csv('scored_customers.csv', index=True)

In [5]:
scored_customers

,customer_id,type_of_customer
0,22853,Vegan People
1,994,Vegan People
2,29221,Vegan People
3,1818,Vegan People
4,7272,Vegan People
...,...,...
1434,25714,Big Family
1435,1030,Big Family
1436,5852,Big Family
1437,14274,Big Family


In [6]:
cluster_1.set_index('customer_id', inplace=True)
cluster_2.set_index('customer_id', inplace=True)
cluster_3.set_index('customer_id', inplace=True)
cluster_4.set_index('customer_id', inplace=True)
cluster_5.set_index('customer_id', inplace=True)
cluster_6.set_index('customer_id', inplace=True)
cluster_7.set_index('customer_id', inplace=True)
cluster_8.set_index('customer_id', inplace=True)
cluster_9.set_index('customer_id', inplace=True)

In [7]:
dataset_basket = pd.read_csv('data/customer_basket.csv').set_index('customer_id')

In [8]:
import ast

# Convert the string representation of lists to actual lists
dataset_basket['list_of_goods'] = dataset_basket['list_of_goods'].apply(ast.literal_eval)

In [9]:
cluster_1_basket = dataset_basket[dataset_basket.index.isin(cluster_1.index)]
cluster_2_basket = dataset_basket[dataset_basket.index.isin(cluster_2.index)]
cluster_3_basket = dataset_basket[dataset_basket.index.isin(cluster_3.index)]
cluster_4_basket = dataset_basket[dataset_basket.index.isin(cluster_4.index)]
cluster_5_basket = dataset_basket[dataset_basket.index.isin(cluster_5.index)]
cluster_6_basket = dataset_basket[dataset_basket.index.isin(cluster_6.index)]
cluster_7_basket = dataset_basket[dataset_basket.index.isin(cluster_7.index)]
cluster_8_basket = dataset_basket[dataset_basket.index.isin(cluster_8.index)]
cluster_9_basket = dataset_basket[dataset_basket.index.isin(cluster_9.index)]
cluster_1_basket

,invoice_id,list_of_goods
customer_id,,
5902,7188908,"[avocado, portal, pet food, ratchet & clank 3,..."
20070,11588714,"[pickles, mashed potato, asparagus, vegetables..."
28051,10500042,"[black beer, shallot, eggplant, flax seed, asp..."
21585,1786347,"[meatballs, megaman zero 3, frozen vegetables,..."
31512,326870,"[vacuum cleaner, tomatoes, herb & pepper, beer]"
...,...,...
30081,1889450,"[deodorant, toothpaste, carrots, melons, tomat..."
27606,3441932,"[yams, tooth brush, shallot, cotton buds, poke..."
23771,6850385,"[shallot, oatmeal, mashed potato, cottage chee..."


----

To understand which promotions and offers can be made for **each cluster of customers** we applied **Apriori algorithm**:

## Cluster 1 - **Vegan People**

In [10]:
te = TransactionEncoder()
te_fit = te.fit(cluster_1_basket['list_of_goods']).transform(cluster_1_basket['list_of_goods'])
dataset_cluster_1_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_1_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,True,True,False,False,False,False,False,...,False,True,False,False,False,False,False,False,False,False
1,False,False,False,True,False,False,False,False,False,False,...,False,True,True,False,False,False,False,False,False,False
2,False,True,False,True,False,False,False,False,False,True,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,True,False,...,True,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13517,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
13518,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,True,False,False
13519,False,False,False,True,True,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
13520,False,False,False,True,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


## Apriori algorithm

'support' will tell us the most popular items that are bought by the customers of this cluster

Below it we can see the item(s) that is(are) bought together with the most popular items calculated by 'support' (sorted by the higher confidence to the lower)

In [11]:
frequent_itemsets = apriori(dataset_cluster_1_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
30,0.782503,(tomatoes)
0,0.657151,(asparagus)
5,0.554652,(carrots)
54,0.537642,"(asparagus, tomatoes)"
76,0.457625,"(carrots, tomatoes)"
...,...,...
58,0.051620,"(mashed potato, avocado)"
120,0.051546,"(asparagus, carrots, shower gel)"
91,0.050880,"(green beans, melons)"
72,0.050362,"(razor, carrots)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
191,"(shallot, asparagus, carrots)",(tomatoes),0.060716,0.782503,0.054208,0.892814,1.140972,1.0,0.006698,2.029152,0.131541,0.068704,0.507183,0.481044
195,"(asparagus, carrots, zucchini)",(tomatoes),0.075063,0.782503,0.066410,0.884729,1.130640,1.0,0.007673,1.886837,0.124923,0.083941,0.470012,0.484799
187,"(asparagus, carrots, melons)",(tomatoes),0.112483,0.782503,0.099246,0.882314,1.127554,1.0,0.011227,1.848120,0.127462,0.124721,0.458910,0.504573
202,"(mashed potato, asparagus, melons)",(tomatoes),0.070774,0.782503,0.061899,0.874608,1.117706,1.0,0.006519,1.734542,0.113332,0.078217,0.423479,0.476856
168,"(cauliflower, carrots, asparagus)",(tomatoes),0.091776,0.782503,0.080240,0.874295,1.117306,1.0,0.008424,1.730220,0.115599,0.101052,0.422039,0.488419
153,"(carrots, toothpaste)",(tomatoes),0.071291,0.782503,0.062269,0.873444,1.116219,1.0,0.006483,1.718586,0.112111,0.078670,0.418126,0.476510
171,"(asparagus, carrots, flax seed)",(tomatoes),0.059163,0.782503,0.051620,0.872500,1.115012,1.0,0.005325,1.705862,0.109635,0.065337,0.413786,0.469234
175,"(asparagus, carrots, frozen vegetables)",(tomatoes),0.097988,0.782503,0.085490,0.872453,1.114952,1.0,0.008814,1.705231,0.114301,0.107535,0.413569,0.490853
160,"(mashed potato, green beans)",(tomatoes),0.065597,0.782503,0.057166,0.871477,1.113705,1.0,0.005836,1.692282,0.109263,0.072277,0.409082,0.472266
116,"(shallot, asparagus)",(tomatoes),0.094661,0.782503,0.082310,0.869531,1.111218,1.0,0.008238,1.667046,0.110552,0.103554,0.400137,0.487360


Associations with more than 1 item:

In [12]:
frequent_itemsets_2 = apriori(
    dataset_cluster_1_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1876,"(cauliflower, tomatoes, frozen vegetables)","(carrots, asparagus)",0.040748,0.389218,0.021594,0.529946,1.361566,1.0,0.005734,1.299387,0.276832,0.052879,0.230406,0.292714
2028,"(carrots, zucchini)","(asparagus, tomatoes, melons)",0.100355,0.153232,0.020929,0.208548,1.360999,1.0,0.005551,1.069892,0.294834,0.089955,0.065327,0.172566
1583,"(shallot, asparagus)","(frozen vegetables, tomatoes)",0.094661,0.184144,0.023591,0.249219,1.353388,1.0,0.006160,1.086676,0.288415,0.092437,0.079762,0.188666
2029,"(zucchini, melons)","(carrots, tomatoes, asparagus)",0.046665,0.331756,0.020929,0.448494,1.351882,1.0,0.005448,1.211673,0.273031,0.058544,0.174695,0.255790
2002,"(shallot, asparagus)","(mashed potato, carrots, tomatoes)",0.094661,0.179855,0.023000,0.242969,1.350914,1.0,0.005974,1.083370,0.286921,0.091444,0.076954,0.185424
1881,"(cauliflower, frozen vegetables)","(carrots, tomatoes, asparagus)",0.048218,0.331756,0.021594,0.447853,1.349948,1.0,0.005598,1.210265,0.272363,0.060256,0.173735,0.256472
2022,"(carrots, zucchini, tomatoes)","(asparagus, melons)",0.086304,0.180003,0.020929,0.242502,1.347212,1.0,0.005394,1.082507,0.282070,0.085292,0.076219,0.179386
2023,"(zucchini, tomatoes, melons)","(carrots, asparagus)",0.040157,0.389218,0.020929,0.521179,1.339042,1.0,0.005299,1.275596,0.263790,0.051240,0.216052,0.287475
1880,"(asparagus, carrots, frozen vegetables)","(cauliflower, tomatoes)",0.097988,0.164695,0.021594,0.220377,1.338097,1.0,0.005456,1.071423,0.280119,0.089571,0.066662,0.175748
2001,"(shallot, carrots)","(mashed potato, asparagus, tomatoes)",0.082828,0.208327,0.023000,0.277679,1.332897,1.0,0.005744,1.096012,0.272309,0.085769,0.087601,0.194040


----

## Cluster 2 - **High Spenders**

In [13]:
te = TransactionEncoder()
te_fit = te.fit(cluster_2_basket['list_of_goods']).transform(cluster_2_basket['list_of_goods'])
dataset_cluster_2_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_2_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
3,False,False,False,False,True,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11900,False,True,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,True,False,False
11901,True,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
11902,True,False,False,False,False,False,False,False,False,False,...,False,True,False,False,False,False,False,False,False,True
11903,False,False,False,False,True,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


## Apriori algorithm

In [14]:
frequent_itemsets = apriori(dataset_cluster_2_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
7,0.762705,(champagne)
12,0.549517,(fresh tuna)
69,0.453591,"(champagne, fresh tuna)"
5,0.429399,(bluetooth headphones)
53,0.356405,"(bluetooth headphones, champagne)"
...,...,...
109,0.051071,"(bluetooth headphones, champagne, almonds)"
152,0.050819,"(cottage cheese, fresh tuna, frozen smoothie)"
133,0.050483,"(bluetooth headphones, salmon, fresh tuna)"
111,0.050315,"(fresh tuna, asparagus, champagne)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
69,"(cottage cheese, frozen smoothie)",(champagne),0.076438,0.762705,0.067955,0.889011,1.165603,1.0,0.009655,2.138007,0.153834,0.088117,0.532275,0.489054
109,"(fresh tuna, frozen smoothie, spaghetti)",(champagne),0.073415,0.762705,0.065183,0.887872,1.164110,1.0,0.009189,2.116286,0.152144,0.084550,0.527474,0.486667
103,"(bluetooth headphones, fresh tuna, frozen smoo...",(champagne),0.088198,0.762705,0.078202,0.886667,1.162529,1.0,0.010933,2.093782,0.153330,0.101207,0.522395,0.494600
101,"(bluetooth headphones, cottage cheese, fresh t...",(champagne),0.069635,0.762705,0.061739,0.886610,1.162456,1.0,0.008628,2.092742,0.150212,0.080118,0.522158,0.483779
92,"(salmon, spaghetti)",(champagne),0.062495,0.762705,0.055271,0.884409,1.159569,1.0,0.007606,2.052880,0.146784,0.071787,0.512879,0.478438
56,"(bluetooth headphones, fromage blanc)",(champagne),0.068123,0.762705,0.060143,0.882861,1.157539,1.0,0.008185,2.025752,0.146048,0.078038,0.506356,0.480858
74,"(fresh tuna, fromage blanc)",(champagne),0.088870,0.762705,0.078370,0.881853,1.156217,1.0,0.010589,2.008467,0.148289,0.101358,0.502108,0.492303
50,"(blueberries, bluetooth headphones)",(champagne),0.061403,0.762705,0.053927,0.878249,1.151493,1.0,0.007095,1.949021,0.140169,0.070019,0.486922,0.474477
90,"(frozen smoothie, spaghetti)",(champagne),0.110458,0.762705,0.096850,0.876806,1.149601,1.0,0.012603,1.926193,0.146292,0.124757,0.480841,0.501894
57,"(bluetooth headphones, frozen smoothie)",(champagne),0.136077,0.762705,0.119026,0.874691,1.146828,1.0,0.015239,1.893686,0.148196,0.152645,0.471929,0.515374


Associations with more than 1 item:

In [15]:
frequent_itemsets_2 = apriori(
    dataset_cluster_2_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1601,"(cottage cheese, spaghetti)","(champagne, frozen smoothie)",0.082990,0.240655,0.028727,0.346154,1.438381,1.0,0.008755,1.161351,0.332356,0.097408,0.138934,0.232763
1731,"(salmon, spaghetti)","(champagne, frozen smoothie)",0.062495,0.240655,0.021588,0.345430,1.435374,1.0,0.006548,1.160067,0.323537,0.076671,0.137981,0.217567
1737,"(cottage cheese, spaghetti)","(fresh tuna, frozen smoothie)",0.082990,0.178916,0.021252,0.256073,1.431243,1.0,0.006403,1.103715,0.328575,0.088307,0.093969,0.187426
1790,"(bluetooth headphones, fresh tuna, frozen smoo...","(cottage cheese, champagne)",0.088198,0.183788,0.023100,0.261905,1.425035,1.0,0.006890,1.105835,0.327114,0.092811,0.095706,0.193795
1602,"(frozen smoothie, spaghetti)","(champagne, cottage cheese)",0.110458,0.183788,0.028727,0.260076,1.415085,1.0,0.008427,1.103102,0.329752,0.108194,0.093466,0.208192
1792,"(bluetooth headphones, cottage cheese)","(fresh tuna, champagne, frozen smoothie)",0.106258,0.154221,0.023100,0.217391,1.409610,1.0,0.006712,1.080718,0.325132,0.097311,0.074689,0.183587
1786,"(bluetooth headphones, cottage cheese, champagne)","(fresh tuna, frozen smoothie)",0.091894,0.178916,0.023100,0.251371,1.404964,1.0,0.006658,1.096783,0.317406,0.093252,0.088243,0.190240
1733,"(fresh tuna, cottage cheese, spaghetti)",(frozen smoothie),0.053171,0.285594,0.021252,0.399684,1.399482,1.0,0.006066,1.190050,0.301480,0.066931,0.159699,0.237048
1570,"(fresh tuna, cottage cheese)","(champagne, frozen smoothie)",0.135321,0.240655,0.045443,0.335816,1.395425,1.0,0.012877,1.143275,0.327720,0.137484,0.125320,0.262323
1228,"(frozen smoothie, avocado)","(champagne, cottage cheese)",0.083998,0.183788,0.021504,0.256000,1.392907,1.0,0.006066,1.097059,0.307944,0.087312,0.088472,0.186501


-----

## Cluster 3 - **Young Famalies with Issues**

In [16]:
te = TransactionEncoder()
te_fit = te.fit(cluster_3_basket['list_of_goods']).transform(cluster_3_basket['list_of_goods'])
dataset_cluster_3_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_3_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,True,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,True,False,False,False,False,True,False,...,False,False,True,False,False,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8595,False,False,False,True,False,False,False,False,True,False,...,False,True,False,False,False,False,False,False,False,False
8596,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
8597,False,False,False,False,False,False,True,False,False,False,...,False,False,False,False,False,False,False,False,False,False
8598,False,False,False,False,False,False,True,False,False,False,...,False,True,False,False,False,False,False,False,False,False


## Apriori algorithm

In [17]:
frequent_itemsets = apriori(dataset_cluster_3_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))


,support,itemsets
26,0.788256,(oil)
9,0.566860,(cooking oil)
85,0.467093,"(cooking oil, oil)"
3,0.430000,(cake)
61,0.358256,"(cake, oil)"
...,...,...
165,0.051163,"(cooking oil, cake, tomatoes, oil)"
107,0.050814,"(salad, oil)"
122,0.050581,"(asparagus, carrots, oil)"
81,0.050233,"(ham, cooking oil)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
99,"(asparagus, cake, cooking oil)",(oil),0.082209,0.788256,0.072442,0.881188,1.117896,1.0,0.007640,1.782180,0.114909,0.090777,0.438890,0.486545
51,"(asparagus, carrots)",(oil),0.057558,0.788256,0.050581,0.878788,1.114851,1.0,0.005211,1.746890,0.109311,0.063606,0.427554,0.471478
101,"(asparagus, cooking oil, candy bars)",(oil),0.066512,0.788256,0.058372,0.877622,1.113373,1.0,0.005944,1.730252,0.109083,0.073295,0.422050,0.475837
61,"(carrots, cake)",(oil),0.078953,0.788256,0.069186,0.876289,1.111681,1.0,0.006951,1.711599,0.109073,0.086697,0.415751,0.482030
113,"(cooking oil, soup, candy bars)",(oil),0.060116,0.788256,0.052674,0.876209,1.111579,1.0,0.005287,1.710496,0.106800,0.066199,0.415374,0.471516
95,"(cooking oil, vegetables mix)",(oil),0.076279,0.788256,0.066512,0.871951,1.106178,1.0,0.006384,1.653621,0.103913,0.083345,0.395267,0.478165
56,"(asparagus, soup)",(oil),0.073721,0.788256,0.064070,0.869085,1.102542,1.0,0.005959,1.617419,0.100407,0.080297,0.381731,0.475183
81,"(cooking oil, carrots)",(oil),0.103372,0.788256,0.089767,0.868391,1.101662,1.0,0.008284,1.608894,0.102920,0.111949,0.378455,0.491136
57,"(asparagus, tomatoes)",(oil),0.064651,0.788256,0.056047,0.866906,1.099778,1.0,0.005085,1.590943,0.096997,0.070334,0.371442,0.469004
107,"(cooking oil, soup, cake)",(oil),0.072209,0.788256,0.062558,0.866345,1.099065,1.0,0.005639,1.584255,0.097151,0.078403,0.368788,0.472854


Associations with more than 1 item:

In [18]:
frequent_itemsets_2 = apriori(
    dataset_cluster_3_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1201,"(mint, oil)","(asparagus, cake)",0.097558,0.125000,0.020698,0.212157,1.697259,1.0,0.008503,1.110628,0.455226,0.102535,0.099608,0.188869
1200,"(mint, cake)","(asparagus, oil)",0.056977,0.221860,0.020698,0.363265,1.637359,1.0,0.008057,1.222078,0.412779,0.080180,0.181722,0.228278
344,(mint),"(asparagus, cake)",0.117907,0.125000,0.023721,0.201183,1.609467,1.0,0.008983,1.095370,0.429293,0.108223,0.087067,0.195475
1198,"(mint, cake, oil)",(asparagus),0.048721,0.267093,0.020698,0.424821,1.590536,1.0,0.007685,1.274224,0.390297,0.070134,0.215209,0.251157
343,"(mint, cake)",(asparagus),0.056977,0.267093,0.023721,0.416327,1.558732,1.0,0.008503,1.255680,0.380110,0.078978,0.203619,0.252569
1195,"(cake, tomatoes)","(asparagus, cooking oil)",0.091279,0.162093,0.022907,0.250955,1.548218,1.0,0.008111,1.118634,0.389665,0.099395,0.106053,0.196138
1327,"(napkins, asparagus)","(tomatoes, oil)",0.087093,0.159884,0.021512,0.246996,1.544848,1.0,0.007587,1.115686,0.386334,0.095410,0.103691,0.190771
1324,"(napkins, asparagus, oil)",(tomatoes),0.074186,0.192442,0.021512,0.289969,1.506786,1.0,0.007235,1.137356,0.363286,0.087761,0.120768,0.200876
1640,"(olive oil, oil)","(cooking oil, candy bars)",0.068372,0.211279,0.021744,0.318027,1.505247,1.0,0.007299,1.156528,0.360291,0.084310,0.135343,0.210472
1328,"(napkins, tomatoes)","(asparagus, oil)",0.065000,0.221860,0.021512,0.330948,1.491695,1.0,0.007091,1.163048,0.352536,0.081069,0.140190,0.213954


---

## Cluster 4 - **Loyal Stable Shoppers**

In [19]:
te = TransactionEncoder()
te_fit = te.fit(cluster_1_basket['list_of_goods']).transform(cluster_4_basket['list_of_goods'])
dataset_cluster_4_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_4_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,True,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16748,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
16749,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
16750,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,True,False
16751,False,False,False,True,False,False,False,False,False,False,...,False,False,False,False,True,False,False,False,False,False


## Apriori algorithm

In [20]:
frequent_itemsets = apriori(dataset_cluster_4_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
21,0.765475,(oil)
7,0.533994,(cooking oil)
31,0.529040,(tea)
75,0.427088,"(cooking oil, oil)"
110,0.425297,"(tea, oil)"
...,...,...
165,0.050558,"(cooking oil, pet food, oil)"
180,0.050439,"(tea, pet food, oil)"
107,0.050319,"(sandwich, oil)"
153,0.050081,"(cooking oil, tea, french fries)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
102,"(tea, cake, candy bars)",(oil),0.087686,0.765475,0.076524,0.872703,1.140080,1.0,0.009402,1.842339,0.134678,0.098532,0.457212,0.486336
108,"(cooking oil, tea, cake)",(oil),0.139557,0.765475,0.120158,0.860992,1.124782,1.0,0.013330,1.687138,0.128932,0.153091,0.407280,0.508982
88,"(milk, tea)",(oil),0.059392,0.765475,0.051095,0.860302,1.123880,1.0,0.005632,1.678795,0.117185,0.066034,0.404335,0.463526
121,"(cooking oil, tea, muffins)",(oil),0.081896,0.765475,0.070375,0.859329,1.122610,1.0,0.007686,1.667194,0.118961,0.090574,0.400190,0.475633
115,"(cooking oil, tea, candy bars)",(oil),0.113233,0.765475,0.097296,0.859251,1.122508,1.0,0.010619,1.666271,0.123074,0.124513,0.399857,0.493178
98,"(cooking oil, cake, candy bars)",(oil),0.083746,0.765475,0.071868,0.858161,1.121083,1.0,0.007762,1.653461,0.117877,0.092452,0.395208,0.476024
113,"(tea, cake, muffins)",(oil),0.064526,0.765475,0.055333,0.857539,1.120271,1.0,0.005941,1.646245,0.114764,0.071429,0.392557,0.464913
45,"(cake, olive oil)",(oil),0.059094,0.765475,0.050618,0.856566,1.118999,1.0,0.005383,1.635070,0.113023,0.065402,0.388405,0.461346
124,"(napkins, cooking oil, tea)",(oil),0.071151,0.765475,0.060885,0.855705,1.117874,1.0,0.006420,1.625314,0.113522,0.078486,0.384734,0.467622
44,"(napkins, cake)",(oil),0.094132,0.765475,0.080404,0.854153,1.115848,1.0,0.008348,1.608027,0.114609,0.103187,0.378120,0.479595


Associations with more than 1 item:

In [21]:
frequent_itemsets_2 = apriori(
    dataset_cluster_4_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1659,"(gums, muffins)","(tea, cake)",0.070972,0.233809,0.022921,0.322960,1.381301,1.0,0.006327,1.131679,0.297133,0.081321,0.116357,0.210497
2229,"(tea, olive oil)","(muffins, oil)",0.077658,0.187787,0.020116,0.259032,1.379388,1.0,0.005533,1.096150,0.298198,0.081995,0.087716,0.183076
1499,"(gums, muffins)","(cooking oil, cake)",0.070972,0.232496,0.022742,0.320437,1.378251,1.0,0.006241,1.129409,0.295408,0.081012,0.114581,0.209128
2398,"(fresh bread, candy bars)","(cooking oil, tea, oil)",0.057900,0.251776,0.020056,0.346392,1.375794,1.0,0.005478,1.144759,0.289934,0.069250,0.126454,0.213025
2324,"(cake, fresh bread)","(cooking oil, tea, oil)",0.069182,0.251776,0.023936,0.345988,1.374191,1.0,0.006518,1.144053,0.292537,0.080587,0.125914,0.220528
1658,"(cake, muffins)","(tea, gums)",0.105235,0.158658,0.022921,0.217811,1.372829,1.0,0.006225,1.075624,0.303518,0.095120,0.070307,0.181140
2394,"(fresh bread, candy bars, oil)","(cooking oil, tea)",0.048469,0.303170,0.020056,0.413793,1.364890,1.0,0.005362,1.188711,0.280958,0.060486,0.158752,0.239974
1500,"(cake, muffins)","(cooking oil, gums)",0.105235,0.159554,0.022742,0.216109,1.354460,1.0,0.005952,1.072147,0.292477,0.093958,0.067292,0.179323
2464,"(gums, muffins)","(cooking oil, tea, oil)",0.070972,0.251776,0.024175,0.340622,1.352880,1.0,0.006306,1.134743,0.280762,0.080968,0.118743,0.218320
2458,"(cooking oil, muffins, oil)","(tea, gums)",0.113114,0.158658,0.024175,0.213720,1.347049,1.0,0.006228,1.070029,0.290496,0.097637,0.065446,0.183045


-----

## Cluster 5 - **Tech-Gamer Customers**

In [22]:
te = TransactionEncoder()
te_fit = te.fit(cluster_5_basket['list_of_goods']).transform(cluster_5_basket['list_of_goods'])
dataset_cluster_5_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_5_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,True,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,True,False,False,False,False,False,False,False,False,True,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False
4,True,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10771,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,True
10772,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,True,False,False,False,False
10773,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
10774,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


## Apriori algorithm

In [23]:
frequent_itemsets = apriori(dataset_cluster_5_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
11,0.585932,(energy drink)
22,0.425019,(pancakes)
73,0.346325,"(energy drink, pancakes)"
28,0.328415,(protein bar)
14,0.297420,(gadget for tiktok streaming)
...,...,...
130,0.051132,"(airpods, gadget for tiktok streaming, pancake..."
80,0.050761,"(french wine, pancakes)"
54,0.050761,"(champagne, cider)"
96,0.050204,"(protein bar, metroid fusion)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
66,"(protein bar, pancakes, energy bar)",(energy drink),0.065794,0.585932,0.058556,0.889986,1.518924,1.0,0.020005,3.763775,0.365700,0.098717,0.734309,0.494961
46,"(pancakes, iPad)",(energy drink),0.062361,0.585932,0.054751,0.877976,1.498428,1.0,0.018212,3.393340,0.354757,0.092245,0.705305,0.485710
67,"(protein bar, gadget for tiktok streaming, pan...",(energy drink),0.078879,0.585932,0.069042,0.875294,1.493850,1.0,0.022825,3.320359,0.358898,0.115888,0.698828,0.496564
34,"(protein bar, energy bar)",(energy drink),0.103842,0.585932,0.090664,0.873101,1.490107,1.0,0.029820,3.262975,0.367019,0.151332,0.693531,0.513918
31,"(energy bar, gadget for tiktok streaming)",(energy drink),0.080271,0.585932,0.070063,0.872832,1.489649,1.0,0.023030,3.256083,0.357389,0.117528,0.692882,0.496204
41,"(half-life 2, pancakes)",(energy drink),0.070341,0.585932,0.061340,0.872032,1.488282,1.0,0.020125,3.235709,0.352908,0.103104,0.690949,0.488360
61,"(airpods, gadget for tiktok streaming, pancakes)",(energy drink),0.058742,0.585932,0.051132,0.870458,1.485597,1.0,0.016714,3.196405,0.347269,0.086148,0.687149,0.478862
33,"(energy bar, pancakes)",(energy drink),0.129918,0.585932,0.112287,0.864286,1.475062,1.0,0.036163,3.051030,0.370152,0.186039,0.672242,0.527962
64,"(airpods, protein bar, pancakes)",(energy drink),0.073589,0.585932,0.063567,0.863808,1.474247,1.0,0.020449,3.040335,0.347241,0.106665,0.671089,0.486149
47,"(iphone 10, pancakes)",(energy drink),0.101800,0.585932,0.087602,0.860529,1.468650,1.0,0.027954,2.968843,0.355269,0.145972,0.663168,0.505019


Associations with more than 1 item:

In [24]:
frequent_itemsets_2 = apriori(
    dataset_cluster_5_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1082,"(white wine, dessert wine)","(champagne, cider)",0.097996,0.050761,0.021158,0.215909,4.253449,1.0,0.016184,1.210624,0.847997,0.165818,0.173980,0.316364
1083,"(champagne, cider)","(white wine, dessert wine)",0.050761,0.097996,0.021158,0.416819,4.253449,1.0,0.016184,1.546697,0.805800,0.165818,0.353461,0.316364
1081,"(white wine, champagne)","(cider, dessert wine)",0.068022,0.074332,0.021158,0.311050,4.184619,1.0,0.016102,1.343594,0.816574,0.174579,0.255727,0.297847
1085,"(cider, dessert wine)","(white wine, champagne)",0.074332,0.068022,0.021158,0.284644,4.184619,1.0,0.016102,1.302818,0.822141,0.174579,0.232433,0.297847
1067,"(beer, cider)","(white wine, dessert wine)",0.063010,0.097996,0.025612,0.406480,4.147945,1.0,0.019438,1.519754,0.809952,0.189171,0.341999,0.333922
1066,"(white wine, dessert wine)","(beer, cider)",0.097996,0.063010,0.025612,0.261364,4.147945,1.0,0.019438,1.268540,0.841367,0.189171,0.211692,0.333922
1084,"(champagne, dessert wine)","(white wine, cider)",0.038512,0.134744,0.021158,0.549398,4.077347,1.0,0.015969,1.920221,0.784973,0.139109,0.479227,0.353211
1069,"(cider, dessert wine)","(white wine, beer)",0.074332,0.084911,0.025612,0.344569,4.058009,1.0,0.019301,1.396164,0.814086,0.191667,0.283752,0.323104
1065,"(white wine, beer)","(cider, dessert wine)",0.084911,0.074332,0.025612,0.301639,4.058009,1.0,0.019301,1.325487,0.823498,0.191667,0.245560,0.323104
1068,"(beer, dessert wine)","(white wine, cider)",0.049740,0.134744,0.025612,0.514925,3.821512,1.0,0.018910,1.783759,0.776970,0.161215,0.439386,0.352504


------

## Cluster 6 - **University Low-Spenders**

In [25]:
te = TransactionEncoder()
te_fit = te.fit(cluster_6_basket['list_of_goods']).transform(cluster_6_basket['list_of_goods'])
dataset_cluster_6_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_6_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,True,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,True,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,False
3,False,False,False,False,True,True,False,True,True,True,...,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1726,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
1727,False,True,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
1728,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
1729,True,True,False,False,False,True,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False


## Apriori algorithm

In [26]:
frequent_itemsets = apriori(dataset_cluster_6_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
2,0.787984,(beer)
37,0.575968,(white wine)
67,0.471404,"(white wine, beer)"
31,0.430387,(salt)
11,0.365685,(cider)
...,...,...
164,0.050260,"(white wine, beer, soup)"
91,0.050260,"(gums, cider)"
176,0.050260,"(white wine, red wine, salt)"
85,0.050260,"(champagne, red wine)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
59,"(white wine, candy bars)",(beer),0.068746,0.787984,0.064125,0.932773,1.183747,1.0,0.009954,3.153741,0.166683,0.080904,0.682916,0.507076
78,"(ketchup, dessert wine)",(beer),0.057192,0.787984,0.053148,0.929293,1.179330,1.0,0.008082,2.998514,0.161285,0.067104,0.666502,0.498371
46,"(salt, black beer)",(beer),0.067013,0.787984,0.061814,0.922414,1.170600,1.0,0.009009,2.732653,0.156205,0.077932,0.634055,0.500430
116,"(white wine, cake, salt)",(beer),0.076256,0.787984,0.069902,0.916667,1.163306,1.0,0.009813,2.544194,0.151970,0.088000,0.606948,0.502688
53,"(champagne, cake)",(beer),0.056615,0.787984,0.050838,0.897959,1.139566,1.0,0.006226,2.077759,0.129822,0.064047,0.518712,0.481238
69,"(ketchup, cider)",(beer),0.063547,0.787984,0.056615,0.890909,1.130619,1.0,0.006541,1.943482,0.123368,0.071221,0.485459,0.481378
19,(oatmeal),(beer),0.078567,0.787984,0.069902,0.889706,1.129092,1.0,0.007992,1.922280,0.124081,0.087745,0.479784,0.489208
56,"(cake, salt)",(beer),0.119584,0.787984,0.106297,0.888889,1.128055,1.0,0.012067,1.908146,0.128937,0.132660,0.475931,0.511893
92,"(white wine, ketchup)",(beer),0.100520,0.787984,0.088966,0.885057,1.123192,1.0,0.009758,1.844541,0.121938,0.111272,0.457860,0.498980
128,"(white wine, salt, french wine)",(beer),0.059503,0.787984,0.052571,0.883495,1.121210,1.0,0.005683,1.819806,0.114946,0.066134,0.450491,0.475105


Associations with more than 1 item:

In [27]:
frequent_itemsets_2 = apriori(
    dataset_cluster_6_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1773,"(beer, muffins)","(salt, dessert wine)",0.079723,0.142692,0.020220,0.253623,1.777416,1.0,0.008844,1.148626,0.475276,0.100000,0.129395,0.197662
1326,"(bramble, white wine)","(beer, red wine)",0.083189,0.153668,0.021375,0.256944,1.672071,1.0,0.008591,1.138988,0.438410,0.099196,0.122028,0.198021
1776,(muffins),"(beer, salt, dessert wine)",0.100520,0.120739,0.020220,0.201149,1.665979,1.0,0.008083,1.100657,0.444426,0.100575,0.091452,0.184307
2065,"(ketchup, salt)","(white wine, beer, cider)",0.067591,0.181398,0.020220,0.299145,1.649110,1.0,0.007959,1.168005,0.422146,0.088384,0.143839,0.205305
1322,"(bramble, beer, white wine)",(red wine),0.069902,0.186020,0.021375,0.305785,1.643832,1.0,0.008372,1.172519,0.421101,0.091133,0.147136,0.210346
1760,"(ketchup, salt)","(beer, dessert wine)",0.067591,0.250722,0.027730,0.410256,1.636299,1.0,0.010783,1.270515,0.417054,0.095427,0.212917,0.260428
1771,"(beer, dessert wine, muffins)",(salt),0.028885,0.430387,0.020220,0.700000,1.626443,1.0,0.007788,1.898710,0.396618,0.046053,0.473327,0.373490
1324,"(white wine, beer, red wine)",(bramble),0.097054,0.135760,0.021375,0.220238,1.622264,1.0,0.008199,1.108339,0.424807,0.101093,0.097749,0.188842
1775,"(dessert wine, muffins)","(beer, salt)",0.034662,0.359908,0.020220,0.583333,1.620787,1.0,0.007744,1.536222,0.396768,0.054012,0.349052,0.319757
1820,"(french fries, salt)","(french wine, beer)",0.086655,0.161179,0.022530,0.260000,1.613118,1.0,0.008563,1.133543,0.416144,0.100000,0.117810,0.199892


------

## Cluster 7 - **Hygiene Obsessed**

In [28]:
te = TransactionEncoder()
te_fit = te.fit(cluster_7_basket['list_of_goods']).transform(cluster_7_basket['list_of_goods'])
dataset_cluster_7_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_7_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,False,True,True,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,True,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,True,False,False,True,False,False,...,False,False,False,True,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15123,False,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,False
15124,True,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
15125,False,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,False
15126,False,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,False


## Apriori algorithm

In [29]:
frequent_itemsets = apriori(dataset_cluster_7_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
26,0.784836,(oil)
2,0.553940,(barbecue sauce)
56,0.465693,"(barbecue sauce, oil)"
9,0.435484,(cologne)
82,0.366142,"(cologne, oil)"
...,...,...
152,0.050436,"(toilet paper, chicken, oil)"
130,0.050370,"(mineral water, barbecue sauce, cologne)"
154,0.050304,"(deodorant, cologne, ham)"
120,0.050238,"(barbecue sauce, candy bars, oil)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
109,"(barbecue sauce, cologne, chicken)",(oil),0.091552,0.784836,0.081703,0.892419,1.137077,1.0,0.009849,2.000014,0.132701,0.102812,0.500003,0.498260
119,"(ham, barbecue sauce, cologne)",(oil),0.080447,0.784836,0.071721,0.891537,1.135953,1.0,0.008584,1.983746,0.130152,0.090379,0.495903,0.491460
113,"(barbecue sauce, cologne, cooking oil)",(oil),0.063062,0.784836,0.055857,0.885744,1.128572,1.0,0.006363,1.883178,0.121593,0.070522,0.468983,0.478457
117,"(barbecue sauce, cologne, fresh tuna)",(oil),0.068020,0.784836,0.060219,0.885326,1.128039,1.0,0.006835,1.876302,0.121790,0.075974,0.467037,0.481027
115,"(deodorant, barbecue sauce, cologne)",(oil),0.108342,0.784836,0.095650,0.882855,1.124891,1.0,0.010620,1.836738,0.124516,0.119934,0.455557,0.502364
82,"(barbecue sauce, samsung galaxy 10)",(oil),0.069672,0.784836,0.061475,0.882353,1.124251,1.0,0.006794,1.828893,0.118796,0.077519,0.453221,0.480341
111,"(deodorant, barbecue sauce, chicken)",(oil),0.071060,0.784836,0.062599,0.880930,1.122439,1.0,0.006828,1.807041,0.117427,0.078910,0.446609,0.480346
121,"(deodorant, barbecue sauce, ham)",(oil),0.064781,0.784836,0.057047,0.880612,1.122033,1.0,0.006204,1.802228,0.116295,0.071977,0.445131,0.476649
100,"(mineral water, cologne)",(oil),0.079389,0.784836,0.069870,0.880100,1.121381,1.0,0.007563,1.794527,0.117576,0.087959,0.442750,0.484563
108,"(ham, fresh tuna)",(oil),0.061740,0.784836,0.054270,0.879015,1.119998,1.0,0.005815,1.778435,0.114192,0.068497,0.437708,0.474082


Associations with more than 1 item:

In [30]:
frequent_itemsets_2 = apriori(
    dataset_cluster_7_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1850,"(cooking oil, cologne)","(barbecue sauce, chicken, oil)",0.097171,0.153622,0.020624,0.212245,1.381601,1.0,0.005696,1.074417,0.305929,0.089604,0.069263,0.173248
1844,"(cooking oil, cologne, oil)","(barbecue sauce, chicken)",0.084545,0.177287,0.020624,0.243941,1.375963,1.0,0.005635,1.088159,0.298471,0.085503,0.081017,0.180136
1857,"(deodorant, cologne, oil)","(barbecue sauce, chicken)",0.144434,0.177287,0.034638,0.239817,1.352703,1.0,0.009031,1.082256,0.304757,0.120654,0.076004,0.217597
1865,"(deodorant, cologne)","(barbecue sauce, chicken, oil)",0.166777,0.153622,0.034638,0.207689,1.351946,1.0,0.009017,1.068239,0.312432,0.121212,0.063880,0.216581
1864,"(barbecue sauce, chicken, oil)","(deodorant, cologne)",0.153622,0.166777,0.034638,0.225473,1.351946,1.0,0.009017,1.075784,0.307576,0.121212,0.070445,0.216581
1866,"(cologne, chicken)","(deodorant, barbecue sauce, oil)",0.140865,0.181980,0.034638,0.245894,1.351211,1.0,0.009003,1.084754,0.302540,0.120183,0.078132,0.218116
1845,"(barbecue sauce, cooking oil, cologne)","(chicken, oil)",0.063062,0.242200,0.020624,0.327044,1.350306,1.0,0.005350,1.126077,0.276888,0.072457,0.111961,0.206098
1860,"(cologne, chicken, oil)","(deodorant, barbecue sauce)",0.122290,0.210074,0.034638,0.283243,1.348302,1.0,0.008948,1.102084,0.294319,0.116341,0.092628,0.224063
1382,"(shrimp, oil)","(barbecue sauce, chicken)",0.092081,0.177287,0.021946,0.238335,1.344342,1.0,0.005621,1.080150,0.282119,0.088699,0.074203,0.181061
1287,"(cooking oil, cologne)","(barbecue sauce, chicken)",0.097171,0.177287,0.023070,0.237415,1.339155,1.0,0.005843,1.078847,0.280519,0.091770,0.073085,0.183771


------

## Cluster 8 - **Loyal Average Spenders**

In [31]:
te = TransactionEncoder()
te_fit = te.fit(cluster_8_basket['list_of_goods']).transform(cluster_8_basket['list_of_goods'])
dataset_cluster_8_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_8_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,True,False,False,False,False,False,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
17142,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,True,False,False,False,False,False
17143,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False
17144,False,False,False,True,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
17145,False,False,True,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


## Apriori algorithm

In [32]:
frequent_itemsets = apriori(dataset_cluster_8_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
1,0.771447,(babies food)
27,0.551583,(ratchet & clank)
60,0.456056,"(babies food, ratchet & clank)"
16,0.425672,(minecraft)
51,0.354173,"(babies food, minecraft)"
...,...,...
146,0.050621,"(pokemon sword, babies food, pet food)"
69,0.050504,"(babies food, yogurt cake)"
126,0.050330,"(milk, cooking oil, babies food)"
70,0.050271,"(bacon, ratchet & clank)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
113,"(cooking oil, pet food, ratchet & clank)",(babies food),0.071208,0.771447,0.063568,0.892711,1.157190,1.0,0.008635,2.130254,0.146252,0.081593,0.530573,0.487556
111,"(cooking oil, ratchet & clank, minecraft)",(babies food),0.103517,0.771447,0.091678,0.885634,1.148017,1.0,0.011820,1.998432,0.143820,0.117043,0.499608,0.502236
108,"(burgers, ratchet & clank, minecraft)",(babies food),0.065026,0.771447,0.057386,0.882511,1.143969,1.0,0.007222,1.945318,0.134603,0.073658,0.485945,0.478449
114,"(cooking oil, ratchet & clank 2, ratchet & clank)",(babies food),0.059777,0.771447,0.052721,0.881951,1.143243,1.0,0.006606,1.936090,0.133261,0.067720,0.483495,0.475146
119,"(pet food, ratchet & clank, minecraft)",(babies food),0.087712,0.771447,0.077215,0.880319,1.141127,1.0,0.009549,1.909688,0.135564,0.098747,0.476354,0.490205
44,"(cooking oil, burgers)",(babies food),0.083746,0.771447,0.073599,0.878830,1.139197,1.0,0.008993,1.886220,0.133357,0.094165,0.469839,0.487117
54,"(cooking oil, pet food)",(babies food),0.112323,0.771447,0.098676,0.878505,1.138775,1.0,0.012025,1.881168,0.137284,0.125687,0.468415,0.503208
120,"(ratchet & clank 2, ratchet & clank, minecraft)",(babies food),0.074474,0.771447,0.065318,0.877056,1.136897,1.0,0.007865,1.858996,0.130102,0.083676,0.462075,0.480862
49,"(ratchet & clank 2, burgers)",(babies food),0.059719,0.771447,0.052312,0.875977,1.135498,1.0,0.006242,1.842822,0.126908,0.067166,0.457354,0.471894
71,"(pokemon sword, minecraft)",(babies food),0.083455,0.771447,0.072957,0.874214,1.133213,1.0,0.008576,1.816997,0.128257,0.093303,0.449641,0.484393


Associations with more than 1 item:

In [33]:
frequent_itemsets_2 = apriori(
    dataset_cluster_8_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1844,"(ratchet & clank 2, babies food, minecraft)","(pet food, ratchet & clank)",0.101126,0.176649,0.024727,0.244521,1.384222,1.0,0.006864,1.089840,0.308800,0.097718,0.082434,0.192251
1840,"(babies food, pet food, minecraft)","(ratchet & clank 2, ratchet & clank)",0.118971,0.151338,0.024727,0.207843,1.373367,1.0,0.006722,1.071330,0.308574,0.100689,0.066581,0.185617
1845,"(ratchet & clank 2, pet food)","(babies food, ratchet & clank, minecraft)",0.081880,0.221030,0.024727,0.301994,1.366305,1.0,0.006629,1.115994,0.292009,0.088889,0.103938,0.206934
1846,"(ratchet & clank 2, minecraft)","(babies food, pet food, ratchet & clank)",0.117863,0.153671,0.024727,0.209797,1.365234,1.0,0.006615,1.071027,0.303269,0.100189,0.066317,0.185354
1762,"(ratchet & clank 2, minecraft)","(pet food, ratchet & clank)",0.117863,0.176649,0.028227,0.239485,1.355714,1.0,0.007406,1.082624,0.297438,0.106001,0.076318,0.199637
1763,"(pet food, minecraft)","(ratchet & clank 2, ratchet & clank)",0.137633,0.151338,0.028227,0.205085,1.355140,1.0,0.007397,1.067613,0.303895,0.108253,0.063331,0.195799
1839,"(ratchet & clank 2, pet food, ratchet & clank)","(babies food, minecraft)",0.051554,0.354173,0.024727,0.479638,1.354249,1.0,0.006468,1.241112,0.275802,0.064901,0.194271,0.274728
1843,"(ratchet & clank 2, ratchet & clank, minecraft)","(babies food, pet food)",0.074474,0.245524,0.024727,0.332028,1.352325,1.0,0.006442,1.129503,0.281497,0.083745,0.114655,0.216370
1842,"(ratchet & clank 2, babies food, pet food)","(ratchet & clank, minecraft)",0.071383,0.256896,0.024727,0.346405,1.348425,1.0,0.006389,1.136949,0.278257,0.081460,0.120453,0.221330
1838,"(pet food, ratchet & clank, minecraft)","(ratchet & clank 2, babies food)",0.087712,0.209483,0.024727,0.281915,1.345767,1.0,0.006353,1.100869,0.281632,0.090753,0.091626,0.199977


------

## Cluster 9 - **Big Family**

In [34]:
te = TransactionEncoder()
te_fit = te.fit(cluster_9_basket['list_of_goods']).transform(cluster_9_basket['list_of_goods'])
dataset_cluster_9_basket = pd.DataFrame(te_fit, columns=te.columns_)

dataset_cluster_9_basket

,airpods,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,beer,black beer,...,vacuum cleaner,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,True,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,True,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3917,False,False,False,False,True,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3918,False,False,False,False,False,False,False,False,True,False,...,False,False,False,False,False,False,False,False,False,False
3919,False,False,False,False,False,False,False,False,True,True,...,False,False,True,True,False,False,False,False,False,False
3920,True,False,False,False,False,True,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False


## Apriori algorithm

In [35]:
frequent_itemsets = apriori(dataset_cluster_9_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))

,support,itemsets
41,0.557114,(white wine)
11,0.393677,(cider)
10,0.369199,(champagne)
81,0.322794,"(white wine, cider)"
14,0.310811,(dessert wine)
...,...,...
126,0.050739,"(white wine, samsung galaxy 10, dessert wine)"
39,0.050484,(trout)
118,0.050484,"(french wine, cider, dessert wine)"
68,0.050229,"(gadget for tiktok streaming, champagne)"


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
53,"(beer, cider, champagne)",(white wine),0.058134,0.557114,0.054309,0.934211,1.676876,1.0,0.021922,6.731871,0.428567,0.096818,0.851453,0.515847
56,"(beer, cider, dessert wine)",(white wine),0.078531,0.557114,0.070372,0.896104,1.608476,1.0,0.026621,4.262780,0.410533,0.124493,0.765411,0.511210
60,"(champagne, cider, dessert wine)",(white wine),0.063488,0.557114,0.056604,0.891566,1.600331,1.0,0.021234,4.084396,0.400560,0.100362,0.755166,0.496584
21,"(beer, cider)",(white wine),0.154258,0.557114,0.136665,0.885950,1.590251,1.0,0.050726,3.883278,0.438867,0.237799,0.742486,0.565630
34,"(bramble, dessert wine)",(white wine),0.092555,0.557114,0.081846,0.884298,1.587284,1.0,0.030282,3.827803,0.407730,0.144140,0.738754,0.515604
37,"(champagne, cider)",(white wine),0.129016,0.557114,0.113463,0.879447,1.578577,1.0,0.041586,3.673778,0.420809,0.198130,0.727801,0.541554
16,"(bramble, beer)",(white wine),0.073687,0.557114,0.064763,0.878893,1.577582,1.0,0.023711,3.656975,0.395243,0.114414,0.726550,0.497570
58,"(bramble, cider, dessert wine)",(white wine),0.060683,0.557114,0.053289,0.878151,1.576251,1.0,0.019482,3.634722,0.389201,0.094399,0.724876,0.486902
45,"(gadget for tiktok streaming, cider)",(white wine),0.065273,0.557114,0.057114,0.875000,1.570595,1.0,0.020749,3.543090,0.388668,0.101037,0.717761,0.488759
50,"(red wine, dessert wine)",(white wine),0.067058,0.557114,0.058644,0.874525,1.569742,1.0,0.021285,3.529669,0.389041,0.103697,0.716687,0.489894


Associations with more than 1 item:

In [36]:
frequent_itemsets_2 = apriori(
    dataset_cluster_9_basket, min_support=0.02, use_colnames=True
    )

rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
203,(energy drink),(pancakes),0.087965,0.072922,0.034421,0.391304,5.366069,1.0,0.028007,1.523057,0.892119,0.272177,0.343426,0.431666
204,(pancakes),(energy drink),0.072922,0.087965,0.034421,0.472028,5.366069,1.0,0.028007,1.727430,0.877643,0.272177,0.421105,0.431666
16,(barbecue sauce),(oil),0.045895,0.089495,0.020908,0.455556,5.090282,1.0,0.016800,1.672356,0.842200,0.182628,0.402041,0.344587
17,(oil),(barbecue sauce),0.089495,0.045895,0.020908,0.233618,5.090282,1.0,0.016800,1.244947,0.882529,0.182628,0.196753,0.344587
1032,"(champagne, spaghetti)","(bluetooth headphones, fresh tuna)",0.077002,0.071137,0.026517,0.344371,4.840941,1.0,0.021039,1.416750,0.859621,0.218029,0.294159,0.358565
1029,"(bluetooth headphones, fresh tuna)","(champagne, spaghetti)",0.071137,0.077002,0.026517,0.372760,4.840941,1.0,0.021039,1.471523,0.854194,0.218029,0.320432,0.358565
1031,"(fresh tuna, champagne)","(bluetooth headphones, spaghetti)",0.117287,0.046915,0.026517,0.226087,4.819093,1.0,0.021015,1.231515,0.897791,0.192593,0.187992,0.395652
1030,"(bluetooth headphones, spaghetti)","(fresh tuna, champagne)",0.046915,0.117287,0.026517,0.565217,4.819093,1.0,0.021015,2.030240,0.831502,0.192593,0.507447,0.395652
205,(protein bar),(energy drink),0.066293,0.087965,0.025752,0.388462,4.416076,1.0,0.019921,1.491377,0.828477,0.200397,0.329479,0.340608
206,(energy drink),(protein bar),0.087965,0.066293,0.025752,0.292754,4.416076,1.0,0.019921,1.320201,0.848164,0.200397,0.242540,0.340608


------

Entire dataset, without any cluster specified:

In [37]:
te = TransactionEncoder()
te_fit = te.fit(dataset_basket['list_of_goods']).transform(dataset_basket['list_of_goods'])
dataset_basket = pd.DataFrame(te_fit, columns=te.columns_)

frequent_itemsets = apriori(dataset_basket, min_support=0.05, use_colnames=True)
display(frequent_itemsets.sort_values(by='support', ascending=False))

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6, num_itemsets=100)
rules = rules.sort_values(by='confidence', ascending=False)

display(rules.head(50).sort_values(by='confidence', ascending=False))



frequent_itemsets_2 = apriori(
    dataset_basket, min_support=0.02, use_colnames=True
    )


rules_2 = association_rules(frequent_itemsets_2,
                                  metric="confidence",
                                  min_threshold=0.2, num_itemsets=100)

rules_2.sort_values(by='lift', ascending=False).head(10)

,support,itemsets
43,0.34900,(oil)
17,0.25052,(cooking oil)
3,0.18067,(babies food)
1,0.15952,(asparagus)
8,0.15134,(cake)
...,...,...
56,0.05175,(strong cheese)
47,0.05157,(pokemon sword)
76,0.05071,"(chicken, oil)"
53,0.05018,(samsung galaxy 10)


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
12,"(cooking oil, cake)",(oil),0.07009,0.34900,0.05754,0.820944,2.352277,1.0,0.033079,3.635744,0.618210,0.159148,0.724953,0.492908
2,(barbecue sauce),(oil),0.10281,0.34900,0.07766,0.755374,2.164395,1.0,0.041779,2.661205,0.599625,0.207564,0.624230,0.488948
11,(tea),(oil),0.10951,0.34900,0.07785,0.710894,2.036946,1.0,0.039631,2.251769,0.571673,0.204513,0.555905,0.466980
3,(cake),(oil),0.15134,0.34900,0.10709,0.707612,2.027542,1.0,0.054272,2.226494,0.597167,0.272320,0.550863,0.507230
4,(candy bars),(oil),0.12091,0.34900,0.08261,0.683235,1.957695,1.0,0.040412,2.055154,0.556479,0.213297,0.513419,0.459970
1,(ratchet & clank),(babies food),0.12179,0.18067,0.08237,0.676328,3.743444,1.0,0.060366,2.531360,0.834500,0.374256,0.604955,0.566121
0,(minecraft),(babies food),0.09558,0.18067,0.06423,0.672003,3.719502,1.0,0.046962,2.497976,0.808415,0.302943,0.599676,0.513756
7,(deodorant),(oil),0.08160,0.34900,0.05317,0.651593,1.867029,1.0,0.024692,1.868505,0.505651,0.140874,0.464813,0.401971
6,(cologne),(oil),0.10567,0.34900,0.06737,0.637551,1.826793,1.0,0.030491,1.796114,0.506069,0.173948,0.443242,0.415294
8,(gums),(oil),0.08729,0.34900,0.05553,0.636155,1.822795,1.0,0.025066,1.789225,0.494562,0.145840,0.441099,0.397634


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
146,(pancakes),(energy drink),0.06762,0.08550,0.03932,0.581485,6.800991,1.0,0.033538,2.185106,0.914823,0.345518,0.542356,0.520684
145,(energy drink),(pancakes),0.08550,0.06762,0.03932,0.459883,6.800991,1.0,0.033538,1.726256,0.932709,0.345518,0.420712,0.520684
108,(cider),(white wine),0.06388,0.08212,0.03524,0.551659,6.717722,1.0,0.029994,2.047283,0.909221,0.318165,0.511548,0.490394
107,(white wine),(cider),0.08212,0.06388,0.03524,0.429128,6.717722,1.0,0.029994,1.639807,0.927289,0.318165,0.390172,0.490394
140,(dessert wine),(white wine),0.05475,0.08212,0.02735,0.499543,6.083090,1.0,0.022854,1.834085,0.884009,0.249726,0.454769,0.416296
139,(white wine),(dessert wine),0.08212,0.05475,0.02735,0.333049,6.083090,1.0,0.022854,1.417271,0.910369,0.249726,0.294419,0.416296
276,"(bluetooth headphones, fresh tuna)",(champagne),0.03762,0.14195,0.03102,0.824561,5.808816,1.0,0.025680,4.890885,0.860209,0.208819,0.795538,0.521545
279,(champagne),"(bluetooth headphones, fresh tuna)",0.14195,0.03762,0.03102,0.218528,5.808816,1.0,0.025680,1.231496,0.964801,0.208819,0.187979,0.521545
334,"(fresh tuna, frozen smoothie)",(champagne),0.02528,0.14195,0.02084,0.824367,5.807447,1.0,0.017252,4.885474,0.849277,0.142359,0.795312,0.485590
147,(protein bar),(energy drink),0.06100,0.08550,0.03018,0.494754,5.786598,1.0,0.024964,1.810010,0.880923,0.259457,0.447517,0.423868


-----

## **Cupons  and Promotions Strategy**

Based on the customer segmentation and association rules analysis, the following targeted coupon and promotion strategies are proposed to enhance customer engagement and sales:

- `Buy 1, Get 1 Free on Oil`: Many clusters frequently purchase oil, making it the top product for promotional offers.

- `10% Discount on Mixed Vegetables`: Customers buying four different vegetables receive a 10% discount, supporting the “Vegan People” cluster.

- `Monthly 20% Discount on Selected Vegetables`: Each month, a different vegetable is offered at 20% off, encouraging repeat purchases and appealing to vegan customers.

- `Student Card Discount on Alcoholic Beverages`: Students receive a 30% discount on alcoholic drinks purchases over €20, targeting the “University Low-Spenders” segment.

- `15% Discount on Combined Purchases of Energy Drinks and Protein Bars`: This promotion supports the “Tech-Gamer Customers” who tend to buy these products together.

- `10% Discount on Ratchet & Clank Bundle`: Customers buying Ratchet & Clank with Ratchet & Clank 2 get a 10% discount, targeting the “Big Family” cluster.

- `Energy Drink and Pancake Packs`: Bundled promotions for energy drinks and pancakes, reflecting popular joint purchases.

- `Loyalty Card Birthday Gift`: Customers holding a loyalty card receive a birthday cake as a special gift.
------

## **Achievement-Based Incentives**

To further motivate and reward customers, the following achievement-based incentives are suggested:

- `5% Discount on Electronics for Every 25 Energy Drinks Purchased`: Customers in the “Tech-Gamer” cluster receive a 5% discount on electronics each time their total energy drink purchases reach 25, accumulated over time.

- `Champagne Gift for Top Monthly Spenders`: Customers spending the most each month receive two bottles of champagne, rewarding the “High Spenders” cluster.

- `10% Discount on Games for Purchasing 20 Baby Foods`: Supports the “Big Family” cluster by incentivizing baby food purchases.

- `15% Discount on All Purchases After Buying 30 Bottles of Oil`: Encourages overall increased shopping frequency across all customer segments.

These tailored promotions leverage the insights from clustering and association rules, targeting specific customer behaviours to maximize engagement and sales growth.